source/content/tools/pytest-api-examples.mkd

```diff
@@ -0,0 +1,321 @@
+
+    # Same generated output files
+    raw_file = tmp_path / "raw_data.csv"
+    raw_file.write_text(raw_collected_data.to_csv())
+    averages_file = tmp_path / "sample_averages.csv"
+    averages_file.write_text(
+        raw_collected_data.groupby(["sample_id"]).agg("mean").to_csv()
+    )
+
+    # Return the files for testing.
+    return {"raw": raw_file, "averages": averages_file}
+
+
+# Both these tests use the computation data as a fixture.
+# Which means if either test fails, the other tests will still run.
+# This can also make the tests more modular and easy to read.
+
+
+def test_raw_data_file(computation_data: typing.Dict[str, pathlib.Path]):
+    raw_data_file = computation_data["raw"]
+    assert raw_data_file.exists()
+    assert raw_data_file.read_text()
+
+
+def test_averages_data_file(computation_data: typing.Dict[str, pathlib.Path]):
+    averages_file = computation_data["averages"]
+    assert averages_file.exists()
+    assert averages_file.read_text()
+~~~
+
+## Use LineMatcher for testing large text
+
+The `LineMatcher` helper class provides methods that can reduce boiler plate
+testing large blocks of text. This provides methods for testing presence and
+absence of lines.
+
+
+~~~ python
+def test_large_text():
+    example_text = textwrap.dedent(
+        """
+        Two roads diverged in a yellow wood,
+        And sorry I could not travel both
+        And be one traveler, long I stood
+        And looked down one as far as I could
+        To where it bent in the undergrowth;
+    """
+    )
+
+    matcher = _pytest.pytester.LineMatcher(example_text.splitlines())
+
+    # Check some lines at random
+    matcher.fnmatch_lines_random(["Two roads diverged in a yellow wood,"])
+
+    # Check lines exist with a regex
+    matcher.re_match_lines_random(["Two roads diverged in a .* wood,"])
+
+    # Check lines don't exist with a regex
+    matcher.no_fnmatch_line("And looked down two as far as I could")
+~~~
+
+## Caching large files or computations
+
+Documentation: [Cache config][cache]
+
+Pytest provides a cache that can be used for expensive operations between test
+runs such as large computations or fetching large data. This can used prevent
+expensive computations from slowing down tests. The cache can be cleared using
+the command line flag: `pytest --cache-clear`.
+
+To access the cache the `pytestconfig` fixture needs to be in arguments to a
+fixture, this will be an instance of [`_pytest.config.Config`][config_class].
+The caveat to using the `get/set` methods is they have to be JSON serialisable,
+so in the examples below I covert `pathlib.Path` objects back and forth to
+strings to serialise into the cache.
+
+[cache]: https://docs.pytest.org/en/stable/cache.html#the-new-config-cache-object
+[config_class]: https://docs.pytest.org/en/latest/reference.html#id35
+
+
+~~~ python
+@pytest.fixture
+def example_data_file(pytestconfig: _pytest.config.Config) -> pathlib.Path:
```

Comment:
Pet peeve of mine: code snippets without the necessary imports. I always get annoyed trying to find out what I need to import to get the snippet to run. Not sure where `_pytest` comes from.

Reply:
Thanks for suggestion, i added the imports to the code blocks.